src/scalarstats.jl

```diff
@@ -260,37 +260,49 @@ variation(x) = ((m, s) = mean_and_std(x); s/m)
 realXcY(x::Real, y::Real) = x*y
 realXcY(x::Complex, y::Complex) = real(x)*real(y) + imag(x)*imag(y)
 
+
 """
-    sem(x)
+    sem(x[, weights::AbstractWeights]; [mean,])
+
+Return the standard error of the mean for a collection `x`. When using no weights, this is
+equal to the (sample) standard deviation divided by the sample size. If weights are used, 
+the variance of the sample mean is calculated as follows:
+
+* `AnalyticWeights`: Not implemented.
+* `FrequencyWeights`: ``\\frac{\\sum_{i=1}^n w_i (x_i - \\bar{x_i})^2}{(\\sum w_i) (\\sum w_i - 1)}``
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
 
-Return the standard error of the mean of collection `x`,
-i.e. `sqrt(var(x, corrected=true) / length(x))`.
+The standard error is then the square root of the above quantities.
+
+# References
+
+Sribney, W. 1997. Probability weights, analytic weights, and summary statistics. Stata FAQ.
 """
-function sem(x)
-    y = iterate(x)
-    if y === nothing
-        T = eltype(x)
-        # Return the NaN of the type that we would get, had this collection
-        # contained any elements (this is consistent with std)
-        return oftype(sqrt((abs2(zero(T)) + abs2(zero(T)))/2), NaN)
+sem(x; mean=nothing) = sqrt(var(x; mean=mean, corrected=true) / length(x))
+
+function sem(x::AbstractArray, weights::UnitWeights; mean=nothing)
+    if length(x) ≠ length(weights) 
+        throw(DimensionMismatch("array and weights do not have the same length"))
     end
-    count = 1
-    value, state = y
-    y = iterate(x, state)
-    # Use Welford algorithm as seen in (among other places)
-    # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
-    M = value / 1
-    S = real(zero(M))
-    while y !== nothing
-        value, state = y
-        y = iterate(x, state)
-        count += 1
-        new_M = M + (value - M) / count
-        S = S + realXcY(value - M, value - new_M)
-        M = new_M
+    return sem(x; mean=mean)
+end
+
+# Weighted methods for the above
+function sem(x::AbstractArray, weights::FrequencyWeights; mean=nothing)
+    return sqrt(var(x, weights; mean=mean, corrected=true) / sum(weights))
+end
+
+function sem(x::AbstractArray, weights::ProbabilityWeights; mean=mean(x, weights))
```

Comment:
For consistency, maybe we could use
```suggestion
function sem(x::AbstractArray, weights::ProbabilityWeights; mean=nothing)
```
and then define and use
```julia
_mean = mean === nothing ? mean(x, weights) : mean
```
in the function body.

Reply:
This isn't visible to users so we probably don't care?

Reply:
It's visible in the sense that this would be the only method that does not accept `nothing`. So eg. if you use compare results or performance of different methods with and without precomputed mean you could not just pass different values for `mean` (`nothing` and some value) but would have to use different calls, namely with and without keyword argument. So the inconsistency can be a bit annoying but I'm not sure how relevant this is in practice.

BTW it would also be inconsistent with `var` which uses `nothing` as default value.

Reply:
Fixed